File: packages/firestore/ios/Plugin/Classes/Constraints/QueryStartAtConstraint.swift
import Foundation
import FirebaseFirestore
import Capacitor

@objc public class QueryStartAtConstraint: NSObject, QueryNonFilterConstraint {
    private var type: String
    private var reference: String

    public init(_ queryConstraint: JSObject) {
        self.type = queryConstraint["type"] as! String
        self.reference = queryConstraint["reference"] as! String
    }

    public func toQuery(_ query: Query, completion: @escaping (Query, Error?) -> Void) {
        Firestore.firestore().document(reference).getDocument { documentSnapshot, error in
            if let error = error {
                completion(query, error)
            } else {
                var newQuery: Query
                switch self.type {
                case "startAt":
                    newQuery = query.start(atDocument: documentSnapshot!)
                default:
                    newQuery = query.start(afterDocument: documentSnapshot!)
                }
                completion(newQuery, nil)
            }
        }
    }
}

File: packages/firestore/android/src/main/java/io/capawesome/capacitorjs/plugins/firebase/firestore/classes/constraints/QueryStartAtConstraint.java
package io.capawesome.capacitorjs.plugins.firebase.firestore.classes.constraints;

import com.getcapacitor.JSObject;
import com.google.android.gms.tasks.Task;
import com.google.android.gms.tasks.Tasks;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.Query;
import io.capawesome.capacitorjs.plugins.firebase.firestore.interfaces.QueryNonFilterConstraint;

public class QueryStartAtConstraint implements QueryNonFilterConstraint {

    private String type;
    private String reference;

    public QueryStartAtConstraint(JSObject queryConstraint) {
        this.type = queryConstraint.getString("type");
        this.reference = queryConstraint.getString("reference");
    }

    public String getReference() {
        return reference;
    }

    public Query toQuery(Query query, com.google.firebase.firestore.FirebaseFirestore firestoreInstance) throws Exception {
        Task<DocumentSnapshot> task = firestoreInstance.document(reference).get();
        DocumentSnapshot documentSnapshot = Tasks.await(task);
        if (type.equals("startAt")) {
            return query.startAt(documentSnapshot);
        } else {
            return query.startAfter(documentSnapshot);
        }
    }
}
